apiserver/facades/controller/firewaller/firewaller.go

```diff
@@ -724,3 +601,125 @@ func mapUnitPortsAndResolveSubnetCIDRs(portRangesByEndpoint network.GroupedPortR
+	canAccess, err := f.accessApplication()
+	if err != nil {
+		return params.ExposeInfoResults{}, err
+	}
+
+	result := params.ExposeInfoResults{
+		Results: make([]params.ExposeInfoResult, len(args.Entities)),
+	}
+
+	for i, entity := range args.Entities {
+		tag, err := names.ParseApplicationTag(entity.Tag)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
+		}
+		application, err := f.getApplication(canAccess, tag)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+
+		if !application.IsExposed() {
+			continue
+		}
+
+		result.Results[i].Exposed = true
+		if exposedEndpoints := application.ExposedEndpoints(); len(exposedEndpoints) != 0 {
+			mappedEndpoints := make(map[string]params.ExposedEndpoint)
+			for endpoint, exposeDetails := range exposedEndpoints {
+				mappedEndpoints[endpoint] = params.ExposedEndpoint{
+					ExposeToSpaces: exposeDetails.ExposeToSpaceIDs,
+					ExposeToCIDRs:  exposeDetails.ExposeToCIDRs,
+				}
+			}
+			result.Results[i].ExposedEndpoints = mappedEndpoints
+		}
+	}
+	return result, nil
+}
+
+// SpaceInfos returns a comprehensive representation of either all spaces or
+// a filtered subset of the known spaces and their associated subnet details.
+func (f *FirewallerAPIV6) SpaceInfos(args params.SpaceInfosParams) (params.SpaceInfos, error) {
+	if !f.authorizer.AuthController() {
+		return params.SpaceInfos{}, apiservererrors.ServerError(apiservererrors.ErrPerm)
+	}
+
+	allSpaceInfos, err := f.getSpaceInfos()
+	if err != nil {
+		return params.SpaceInfos{}, apiservererrors.ServerError(err)
+	}
+
+	// Apply filtering if required
+	if len(args.FilterBySpaceIDs) != 0 {
+		var (
+			filteredList network.SpaceInfos
+			selectList   = set.NewStrings(args.FilterBySpaceIDs...)
+		)
+		for _, si := range allSpaceInfos {
+			if selectList.Contains(si.ID) {
+				filteredList = append(filteredList, si)
+			}
+		}
+
+		allSpaceInfos = filteredList
+	}
+
+	return params.FromNetworkSpaceInfos(allSpaceInfos), nil
+}
+
+// WatchSubnets returns a new StringsWatcher that watches the specified
+// subnet tags or all tags if no entities are specified.
+func (f *FirewallerAPIV6) WatchSubnets(args params.Entities) (params.StringsWatchResult, error) {
+	if !f.authorizer.AuthController() {
+		return params.StringsWatchResult{}, apiservererrors.ServerError(apiservererrors.ErrPerm)
+	}
+
+	var (
+		filterFn  func(id interface{}) bool
+		filterSet set.Strings
+		result    = params.StringsWatchResult{}
+	)
+
+	if len(args.Entities) != 0 {
+		filterSet = set.NewStrings()
+		for _, arg := range args.Entities {
+			subnetTag, err := names.ParseSubnetTag(arg.Tag)
+			if err != nil {
+				return params.StringsWatchResult{}, apiservererrors.ServerError(err)
+			}
+
+			filterSet.Add(subnetTag.Id())
+		}
+
+		filterFn = func(id interface{}) bool {
+			return filterSet.Contains(id.(string))
+		}
+	}
```

Comment:
Not for this PR but I wish we had a proper server side business logic layer where we'd have a WatchSubnets(ids ...[]string) and pass in ids extracted from parsed tags and do all the filter setup etc in that business logic layer.